test_bandwidth.py

```diff
@@ -0,0 +1,204 @@
+        busbw = algbw * op.ranks_factor()
+        busbw_total += busbw
+        pretty_table.add_row([(tensor_elements * (torch.finfo(dtype).bits // 8)), tensor_elements,
+                              dtype, collective_op, time_per_node*1000, algbw / 1024**3, busbw / 1024**3])
+    if dist.get_rank() == 0:
+        avg_busbw = busbw_total / len(tensor_list)
+        print(pretty_table)
+        print(f'Average busbw: {avg_busbw/1024**3:.3f} GB/s')    
+
+def get_op(ops: str, world_size: int = 1)-> CollectiveOps:
+    if ops == 'allreduce':
+        return AllReduce(world_size)
+    elif ops == 'allgather':
+        return AllGather(world_size)
+    elif ops == 'broadcast':
+        return Broadcast(world_size)
+    elif ops == 'alltoall':
+        return AllToAll(world_size)
+
+def str_to_int(bytes: str) -> int:
+    if bytes[-1] == 'B':
+        bytes = bytes[:-1]
+    elif bytes[-1] == 'K':
+        return int(bytes[:-1]) * 1024
+    elif bytes[-1] == 'M':
+        return int(bytes[:-1]) * 1024**2
+    elif bytes[-1] == 'G':
+        return int(bytes[:-1]) * 1024**3
+    return int(bytes)
+
+def get_tensor_list(minbytes:str, maxbytes:str, stepbytes:str, stepfactor:str, dtype = torch.float) -> List[int]:
+    minbytes = str_to_int(minbytes) 
+    maxbytes = str_to_int(maxbytes) 
+    stepbytes = str_to_int(stepbytes)  
+    bytes_list = []
+    if stepfactor == 1:
+        bytes_list = list(range(minbytes, maxbytes + 1, stepbytes))
+    else:
+        while minbytes <= maxbytes:
+            bytes_list.append(minbytes)
+            minbytes *= stepfactor
+    # The list of elements number for testing tensors, arranged in ascending order. 
+    # 32M bytes -> 32*1024**2 / (torch.finfo(float).bits // 8)
+    bytes_list = sorted(bytes_list) 
+    tensor_list = [bytes // (torch.finfo(dtype).bits // 8) for bytes in bytes_list]
+    
+    return tensor_list
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-c', '--collective_op', type=str, default='allreduce',
+                        choices=['allreduce', 'allgather', 'broadcast', 'alltoall'])
+    parser.add_argument('-min', '--minbytes', type=str, default='1M')
+    parser.add_argument('-max', '--maxbytes', type=str, default='64M')
+    parser.add_argument('-b', '--stepbytes', type=str, default='1M')
+    parser.add_argument('-f', '--stepfactor', type=int, default=2)
+    parser.add_argument('-w', '--warmup_iters', type=int, default=5) 
+    parser.add_argument('-i', '--iters', type=int, default=20)
+    parser.add_argument('-d', '--dtype', type=str,default='float', 
+                        choices=['fp16', 'bf16', 'float'])
+    args = parser.parse_args()
+
+    #setup
+    if torch.cuda.is_available():
+        print("CUDA is available. GPU can be used.")
+    else:
+        print("CUDA is not available. Using CPU.")
+        sys.exit("Exiting the program due to lack of GPU.")
```

Comment:
This looks weird to me, if you are telling the user that CPU will be used, why does this program exit?

Reply:
Okay, I'll remove it. So it will be the bandwidth of CPUs.